### ISSUE TITLE
Strange inference results, even if the training results are good

### ISSUE THREAD

**Issue by williamhyin** [google/automl#715]
HI all,

**I've had strange inference results， even if the training results are good and I've tried to train three times.**
The following is problem description:
![](https://williamhyin-1301408646.cos.ap-shanghai.myqcloud.com/img/20200831085125.jpg)
I am trying to train a traffic sign detection network which have 200 classes.
This is my  config.  I have modified the num_classes= 200+1, besides that I also add a 0 index for background.

------------------------------------------------------------------------------------------------------
```
num_classes: 201
label_map: {0: background, 1: I-1.1, 2: I-10, 3: I-11, 4: I-13, 5: I-13.1, 6: I-14, 7: I-15, 8: I-16, 9: I-17, 10: I-18, 11: I-19, 12: I-2, 13: I-2.1, 14: I-20, 15: I-25, 16: I-27, 17: I-28, 18: I-28.1, 19: I-29, 20: I-29.1, 21: I-3, 22: I-30, 23: I-32, 24: I-34, 25: I-36, 26: I-37, 27: I-38, 28: I-39-1, 29: I-39-2, 30: I-39-3, 31: I-4, 32: I-5, 33: I-5.1, 34: I-5.2, 35: I-8, 36: I-9, 37: II-1, 38: II-10.1, 39: II-14, 40: II-17, 41: II-18, 42: II-19-4, 43: II-2, 44: II-21, 45: II-22, 46: II-23, 47: II-26, 48: II-26.1, 49: II-28, 50: II-3, 51: II-30-10, 52: II-30-30, 53: II-30-40, 54: II-30-50, 55: II-30-60, 56: II-30-70, 57: II-32, 58: II-33, 59: II-34, 60: II-35, 61: II-39, 62: II-4, 63: II-40, 64: II-41, 65: II-42, 66: II-42.1, 67: II-43, 68: II-45, 69: II-45.1, 70: II-45.2, 71: II-46, 72: II-46.1, 73: II-46.2, 74: II-47, 75: II-47.1, 76: II-48, 77: II-6, 78: II-7, 79: II-7.1, 80: II-8, 81: III-1, 82: III-10, 83: III-105, 84: III-105.1, 85: III-105.3, 86: III-107-1, 87: III-107-2, 88: III-107.1-1, 89: III-107.1-2, 90: III-107.2-1, 91: III-107.2-2, 92: III-112, 93: III-113, 94: III-12, 95: III-120, 96: III-120-1, 97: III-120.1, 98: III-123, 99: III-124, 100: III-14, 101: III-14.1, 102: III-15, 103: III-16, 104: III-18-40, 105: III-18-50, 106: III-18-60, 107: III-18-70, 108: III-2, 109: III-202-5, 110: III-203-2, 111: III-206-1, 112: III-21, 113: III-23, 114: III-25, 115: III-25.1, 116: III-27, 117: III-29-30, 118: III-29-40, 119: III-3, 120: III-30-30, 121: III-33, 122: III-34, 123: III-35, 124: III-37, 125: III-39, 126: III-40, 127: III-42, 128: III-43, 129: III-45, 130: III-46, 131: III-47, 132: III-5, 133: III-50, 134: III-54, 135: III-59, 136: III-6, 137: III-64, 138: III-68, 139: III-74, 140: III-77, 141: III-78, 142: III-8-1, 143: III-84, 144: III-84-1, 145: III-85-2, 146: III-85-3, 147: III-85.1, 148: III-86-1, 149: III-86-2, 150: III-87, 151: III-90, 152: III-90.1, 153: III-90.2, 154: III-91, 155: IV-1, 156: IV-1.1, 157: IV-10, 158: IV-11, 159: IV-12, 160: IV-12.1, 161: IV-13-1, 162: IV-13-2, 163: IV-13-3, 164: IV-13-4, 165: IV-13-5, 166: IV-13-6, 167: IV-13.1-2, 168: IV-13.1-3, 169: IV-13.1-4, 170: IV-16, 171: IV-17, 172: IV-18, 173: IV-2, 174: IV-20-1, 175: IV-3-1, 176: IV-3-2, 177: IV-3-4, 178: IV-3-5, 179: IV-5, 180: IV-6, 181: VI-2.1, 182: VI-3-1, 183: VI-3-2, 184: VI-3.1-1, 185: VI-3.1-2, 186: VI-8, 187: VII-4, 188: VII-4-1, 189: VII-4-2, 190: VII-4.1-1, 191: VII-4.3, 192: VII-4.3-1, 193: VII-4.3-2, 194: VII-4.4-1, 195: VII-4.4-2, 196: X-1.1, 197: X-1.2, 198: X-4, 199: X-6-3, 200: I-1}
moving_average_decay: 0
```
----------------------------------------------------------------------------------------------------------------
I used the EfficientDet D1 with batch_size=4, epochs=50.
This is my training command.

```
d="d1"
times="2"
!mkdir /home/code/sb/automl/efficientdet/tmp/finetune{d}_{times}
!python main.py --mode=train \
    --training_file_pattern=/home/code/sb/automl/efficientdet/dfg/tfrecord/dfg_train* \
    --validation_file_pattern=/home/code/sb/automl/efficientdet/dfg/tfrecord/dfg_val* \
    --model_name=efficientdet-{d} \
    --model_dir=/home/code/sb/automl/efficientdet/tmp/finetune{d}_{times}  \
    --ckpt=/home/code/sb/automl/efficientdet/weights/efficientdet-{d}  \
    --train_batch_size=4 \
    --eval_batch_size=4 --eval_samples=1703 \
    --num_examples_per_epoch=14029 --num_epochs=50  \
    --hparams=/home/code/sb/automl/efficientdet/dfg/dfg.yaml

```
-----------------------------------------------------------------------------------------------------------------------
After 50 epochs, the det-loss and cls_loss decreased to about 0.2.
So i want to test the trained model.
The following is save and inference code: I use the same config .
```
d="d1"
times="2"
!rm  -rf /home/code/sb/automl/efficientdet/tmp/finetune{d}_{times}_export
!python model_inspect.py --runmode=saved_model --model_name=efficientdet-{d} \
  --ckpt_path=/home/code/sb/automl/efficientdet/tmp/finetune{d}_{times} \
  --saved_model_dir=/home/code/sb/automl/efficientdet/tmp/finetune{d}_{times}_export \
  --tflite_path=/home/code/sb/automl/efficientdet/tmp/finetune{d}_{times}_export/efficientdet-{d}-dfg.tflite --hparams=/home/code/sb/automl/efficientdet/dfg/dfg.yaml


d="d1"
times="2"
!python model_inspect.py --runmode=saved_model_infer \
  --model_name=efficientdet-{d}  \
  --saved_model_dir=/home/code/sb/automl/efficientdet/tmp/finetune{d}_{times}_export/\
  --input_image=/home/code/sb/automl/efficientdet/testdata/*.jpg\
  --output_image_dir=/home/code/sb/automl/efficientdet/serve_image_out/ \
  --hparams=/home/code/sb/automl/efficientdet/dfg/dfg.yaml  \

```
----------------------------------------------------------------------------------------------------------------
Actually the results is really strange.
![](https://williamhyin-1301408646.cos.ap-shanghai.myqcloud.com/img/20200831090202.jpg)
There are so many bounding box which 100% confidence.
And most of them exist on the top of image.
Even if the training is not converage, the result should not be so strange.

I have tried many times. freeze the network or not. The results is same . I dont know why..
Another questions, I am training with NVIDIA2080Ti, 128GB RAM, why i can maximum set batch_size=4 when traing efficient D1. Is Efficient det  so computationally expensive?
**Can someone help me ?**

Thanks!

**Comment by NikZak** [comment]
I think this could the same problem as in #676

**Comment by williamhyin** [comment]
> I think this could the same problem as in #676

Yes, I have checked the problem  #676, but actually i did not find the solution from  #676.
Do you have any suggestions for that?
Thanks

**Comment by NikZak** [comment]
Try option "use_keras_model: False" in your config file

**Comment by williamhyin** [comment]
> Try option "use_keras_model: False" in your config file

When i only change config with option "use_keras_model: False" for inference , it did not work. I am traing 10 epochs again, when it functioniert, i will inform you, thank you.

**Comment by williamhyin** [comment]
> Try option "use_keras_model: False" in your config file

Hi

**I'm sure the problem is keras-modul bug.**
It is useful to set "use_keras_model: False" during training.

I did a comparative test with a small data set.
The difference is whether keras is used or not.
The following are results.
without keras:
![](https://williamhyin-1301408646.cos.ap-shanghai.myqcloud.com/img/20200831153556.jpg)

with keras

![](https://williamhyin-1301408646.cos.ap-shanghai.myqcloud.com/img/20200831153637.jpg)

Thank you!

**Comment by LaurensHagendoorn** [comment]
The only meaningfull difference appears to be that the output is a dict?
```
for i in range(params['min_level'], params['max_level'] + 1):
        cls_outputs[i] = cls_out_list[i - params['min_level']]
        box_outputs[i] = box_out_list[i - params['min_level']]
      return cls_outputs, box_outputs
```
I'm guessing the visualisation script included does nothing to take this into account? It doesn't mention "use_keras_model" anyway?

**Comment by LaurensHagendoorn** [comment]
I just fixed some weird results, though it probably has a different cause:
This efficientDet repo takes input in [ymin, xmin, ymax, xmax] format, and also returns it that way, and I had my stuff the other way round.

**Comment by fsx950223** [comment]
Fixed in latest commit.
